ocaml/xapi/xapi_host.ml

```diff
@@ -2671,44 +2671,165 @@ let with_temp_file_contents ~contents f =
-        if contents <> "" then (
-          with_temp_file_contents ~contents
-            (Tar_unix.Archive.extract extract_certificate_file) ;
-          debug "UEFI tar file extracted to temporary directory"
+        else
+          []
+      in
+      files_of_dir !Xapi_globs.default_auth_dir
+      |> List.iter (fun file ->
+             let src = !Xapi_globs.default_auth_dir // file in
+             let dest = !Xapi_globs.varstore_dir // file in
+             let finally = Xapi_stdext_pervasives.Pervasiveext.finally in
+             let src_fd = Unix.openfile src [Unix.O_RDONLY] 0o400 in
+             finally
+               (fun () ->
+                 let dest_fd =
+                   Unix.openfile dest
+                     [Unix.O_WRONLY; Unix.O_CREAT; Unix.O_TRUNC]
+                     0o644
+                 in
+                 finally
+                   (fun () ->
+                     debug "override_uefi_certs: copy_file %s->%s" src dest ;
+                     ignore (Unixext.copy_file src_fd dest_fd)
+                   )
+                   (fun () -> Unix.close dest_fd)
+               )
+               (fun () -> Unix.close src_fd)
+         )
+  | base64_value -> (
+      if
+        Sys.file_exists !Xapi_globs.varstore_dir
+        && Sys.is_directory !Xapi_globs.varstore_dir
+      then
+        match Base64.decode base64_value with
+        | Ok contents ->
+            (* No uefi certificates, nothing to do. *)
+            if contents <> "" then (
+              with_temp_file_contents ~contents
+                (Tar_unix.Archive.extract extract_certificate_file) ;
+              debug "UEFI tar file extracted to temporary directory"
+            )
+        (* No UEFI tar file. *)
+        | Error _ ->
+            debug
+              "UEFI tar file was not extracted: it was not base64-encoded \
+               correctly"
+    )
+
+let write_uefi_certificates_to_disk ~__context ~host =
+  let ( // ) = Filename.concat in
+  let with_valid_symlink from_path to_path fn =
+    debug "override_uefi_certs: with_valid_symlink %s->%s" from_path to_path ;
+    let is_valid_symlink from_path to_path =
+      if Sys.file_exists from_path then
+        Unix.(
+          match (lstat from_path).st_kind with
+          | S_LNK ->
+              readlink from_path = to_path
+          | _ ->
+              false
         )
-    (* No UEFI tar file. *)
-    | Error _ ->
-        debug
-          "UEFI tar file was not extracted: it was not base64-encoded correctly"
+      else
+        false
+    in
+    if not (is_valid_symlink from_path to_path) then (
+      try
+        Helpers.rmtree from_path ;
+        Unix.symlink to_path from_path
+      with e ->
+        let msg = Printf.sprintf "Failed to symlink %s->%s" from_path to_path in
+        error "%s: %s" msg (Printexc.to_string e) ;
+        raise Api_errors.(Server_error (internal_error, [msg]))
+    ) ;
+    fn from_path
+  in
```

Comment:
There is a lot of repeated code around file and directory handling.

Reply:
I've refactored the code to use Helpers.FileSys.* more extensively, the number of line changes reduced considerably since the last patchset.